### PR TITLE
Separate local-only builds from remote revisions

### PR DESCRIPTION
**Pull request by TimoWilken**
This prevents interference between the remote store and local builds in the case where another machine builds a package and uploads it to the remote store with a revision number that conflicts with a local package, in case that package wasn't uploaded to the remote store. In that case, local revision numbers become a mess, and packages could start depending on the wrong revision numbers of dependencies.

It also prevents builds intended only for local use from being uploaded to the remote. This is important as local-only builds might otherwise be uploaded to the remote much later, when some of their dependencies might conflict with new ones on the remote, and would be renamed. In that case, we currently don't update the main package, so it would depend on (and e.g. try to source scripts from) an incorrect revision number of its dependency.

With these changes, if running aliBuild with a read-write remote when some local revisions are left over in the local store from a previous run, then those revisions are not uploaded, and may be deleted from the local store. Running aliBuild repeatedly with a read-only remote preserves local revisions.

### CODE REVIEW

**Comment by TimoWilken** [comment]
TODO: create a new alibuild release and get the CI builders to use it when merged.

**Comment by TimoWilken** [comment]
I think I've figured out why aliBuild with the local-revision patch rebuilt some packages: I built using the HTTPS remote store first, using a clean `sw/` directory. Then I built with no remote store, but only deleting O2 from the store, keeping the rest of `sw/` from the previous run to save time. During the second run, lots of packages suddenly depended on GCC-Toolchain, which changed their hash -- presumably because this package was built during the previous run and was now available in the store, where those packages had used the system GCC before.

**Comment by TimoWilken** [comment]
@ktf: ready to merge in my opinion, I can make a new alibuild release and deploy it for the CI builders gradually.

File: alibuild_helpers/build.py [hunk 1]
```diff
@@ -67,6 +67,7 @@ def getDirectoryHash(d):
 
 # Helper class which does not do anything to sync
 class NoRemoteSync:
+  writeStore = ""
```
Comment (ktf) [inline alibuild_helpers/build.py:70]: This is just some cleanup, right? Can you do it in a separate PR?

File: alibuild_helpers/build.py [hunk 2]
```diff
@@ -792,51 +792,60 @@ def downloadTask(p):
...
       else:
-        busyRevisions.append(revision)
-
-    if not "revision" in spec and busyRevisions:
-      spec["revision"] = min(set(range(1, max(busyRevisions)+2)) - set(busyRevisions))
-    elif not "revision" in spec:
-      spec["revision"] = "1"
+        debug("Package %s with hash %s is already found in %s. Not building.",
+              p, h, d)
+        getstatusoutput("ln -snf %s-%s %s/%s/%s/latest-%s" % (
```
Comment (ktf) [inline alibuild_helpers/build.py:829]: Mmm... No, please do not change the key based formatting. If you want to avoid the `format` helper, simply use str.format with the actual name.

File: alibuild_helpers/cmd.py [hunk 3]
```diff
@@ -10,10 +10,9 @@
 
 def execute(command, printer=debug):
   popen = subprocess.Popen(command, shell=is_string(command), stdout=subprocess.PIPE)
-  lines_iterator = iter(popen.stdout.readline, "")
-  for line in lines_iterator:
+  for line in iter(popen.stdout.readline, ""):
     if not line: break
-    printer(to_unicode(line).strip("\n")) # yield line
+    printer("%s", to_unicode(line).strip("\n"))
```
Comment (ktf) [inline alibuild_helpers/cmd.py:15]: Why doing the formatting for "%s"?

File: alibuild_helpers/cmd.py [hunk 4]
```diff
@@ -10,10 +10,9 @@
 
 def execute(command, printer=debug):
   popen = subprocess.Popen(command, shell=is_string(command), stdout=subprocess.PIPE)
-  lines_iterator = iter(popen.stdout.readline, "")
-  for line in lines_iterator:
+  for line in iter(popen.stdout.readline, ""):
```
Comment (ktf) [inline alibuild_helpers/cmd.py:13]: Also this is unrelated to the PR. Can you please split it off?

File: tests/test_cmd.py [hunk 5]
```diff
@@ -16,7 +16,7 @@ class CmdTestCase(unittest.TestCase):
     def test_execute(self, mock_debug):
       err = execute("echo foo", mock_debug)
       self.assertEqual(err, 0)
-      self.assertEqual(mock_debug.mock_calls, [call('foo')])
+      self.assertEqual(mock_debug.mock_calls, [call("%s", "foo")])
```
Comment (ktf) [inline tests/test_cmd.py:19]: Why the extra formatting?

File: alibuild_helpers/build.py [hunk 6]
```diff
@@ -1205,7 +1215,11 @@ def downloadTask(p):
 
     dieOnError(err, buildErrMsg)
 
-    syncHelper.syncToRemote(p, spec)
+    # Make sure not to upload local-only packages! These might have been
+    # produced in a previous run with a read-only remote store.
+    if not spec["revision"].startswith("local"):
```
Comment (ktf) [inline alibuild_helpers/build.py:1220]: We should stop at the first local package, no?

**Review by ktf — CHANGES_REQUESTED** [review]

File: alibuild_helpers/build.py (hunk 2 above)
Comment (TimoWilken) [reply alibuild_helpers/build.py:829]: Sorry, that's a holdover from some cleanup I was doing before. I'll change that.

File: alibuild_helpers/cmd.py (hunk 3 above)
Comment (TimoWilken) [reply alibuild_helpers/cmd.py:15]: Because `line` can contain `%` signs (e.g. `[100%] Compiling xyz.cxx`), which confuses the string formatting.

File: tests/test_cmd.py (hunk 5 above)
Comment (TimoWilken) [reply tests/test_cmd.py:19]: This is changed because of the change above. In the general case, an output line can contain `%`, so this needs the extra `%s`.

File: alibuild_helpers/build.py (hunk 6 above)
Comment (TimoWilken) [reply alibuild_helpers/build.py:1220]: What should happen in the following case:
```
 D-1        E-local1
  |          |
 B-1        C-local1
   \        /
    A-local1
```
(i.e. A depends partly on local and partly on remote packages, but the remote packages are consistent; `buildOrder` is something like `[D, E, B, C, A]`)

I suppose `B-1` and `D-1` should've been uploaded in a previous alibuild run, so it might be alright to stop uploading when we hit `E-local1`. But if that previous upload failed, would we want it to happen now to make local and remote consistent again?

File: alibuild_helpers/build.py (hunk 1 above)
Comment (TimoWilken) [reply alibuild_helpers/build.py]: I've undone this.

File: alibuild_helpers/cmd.py (hunk 4 above)
Comment (TimoWilken) [reply alibuild_helpers/cmd.py]: Undone.

File: alibuild_helpers/build.py (hunk 6 above)
Comment (ktf) [reply alibuild_helpers/build.py:1219]: I assume that if B-1 dependend on E, it would actually be called B-local1, no? If that is the case, you are right, we should not stop.

Comment (TimoWilken) [reply alibuild_helpers/build.py:1219]: Yes, if B depended on E, it would get a local revision number. When we have a read-only remote, no new packages with a non-local revision are produced, only existing ones reused.

File: alibuild_helpers/build.py [hunk 7]
```diff
@@ -1205,7 +1214,11 @@ def downloadTask(p):
...
+      syncHelper.syncToRemote(p, spec)
```
Comment (ktf) [inline alibuild_helpers/build.py:1220]: Can we invert the logic here? Something like:

```
if spec["revision"].startswith("local"):
  continue       # break?
```

File: alibuild_helpers/build.py [hunk 8]
```diff
@@ -878,6 +886,7 @@ def downloadTask(p):
       if spec["devel_hash"]+spec["deps_hash"] == spec["old_devel_hash"]:
         info("Development package %s does not need rebuild", spec["package"])
         buildOrder.pop(0)
+        packageIterations = 0
```
Comment (ktf) [inline alibuild_helpers/build.py:889]: What is this for?

File: alibuild_helpers/build.py [hunk 9]
```diff
@@ -792,51 +791,60 @@ def downloadTask(p):
...
-    elif not "revision" in spec:
-      spec["revision"] = "1"
+        debug("Package %s with hash %s is already found in %s. Not building.",
+              p, h, d)
+        getstatusoutput("ln -snf {ver}-{rev} {wd}/{arch}/{pkg}/latest-{family}".format(
+          ver=spec["version"], rev=spec["revision"], wd=workDir,
+          arch=args.architecture, pkg=spec["package"], family=spec["build_family"]))
+        getstatusoutput("ln -snf {ver}-{rev} {wd}/{arch}/{pkg}/latest".format(
+          ver=spec["version"], rev=spec["revision"], wd=workDir,
+          arch=args.architecture, pkg=spec["package"]))
```
Comment (ktf) [inline alibuild_helpers/build.py:833]: Did you change this wrt the current code or is it just a nicer version? Can you do a separate PR for this?

File: alibuild_helpers/build.py (hunk 8 above)
Comment (TimoWilken) [reply alibuild_helpers/build.py:889]: I think this was forgotten at some point -- this tracks the number of times the current package was rebuilt, but we've just decided we're done with the current package, so `packageIterations` should be reset.

As a side note, this was/is really unlikely to actually cause a bug because the limit of iterations is set at 20, so one or two more for the next package still won't trip the limit. (Which is why it wasn't noticed before, I presume.)

File: alibuild_helpers/build.py [hunk 10]
```diff
@@ -792,51 +791,60 @@ def downloadTask(p):
...
-      realPath = readlink(d)
-      matcher = format("../../%(a)s/store/[0-9a-f]{2}/([0-9a-f]*)/%(p)s-%(v)s-([0-9]*).%(a)s.tar.gz$",
-                       a=args.architecture,
-                       p=spec["package"],
-                       v=spec["version"])
-      m = re.match(matcher, realPath)
-      if not m:
+      match = re.match(
+        "../../%s/store/[0-9a-f]{2}/([0-9a-f]+)/%s-%s-((?:local)?[0-9]+).%s.tar.gz$" %
+        tuple(map(re.escape, (args.architecture, spec["package"],
```
Comment (ktf) [inline alibuild_helpers/build.py:801]: `[re.escape(x) for x in [args.architecture, spec["package"], spec["version"], args.architecture]]` but then again, why not simply using `str.format`?

File: alibuild_helpers/build.py (hunk 9 above)
Comment (TimoWilken) [reply alibuild_helpers/build.py:833]: No change, just "nicer". I can make a separate PR.

File: alibuild_helpers/build.py [hunk 11]
```diff
@@ -727,7 +727,7 @@ def downloadTask(p):
       logger_handler.setFormatter(
           LogFormatter("%%(asctime)s:%%(levelname)s:%s:%s:%s: %%(message)s" %
                        (mainPackage, p, args.develPrefix if "develPrefix" in args else mainHash[0:8])))
-    if spec["package"] in develPkgs and getattr(syncHelper, "writeStore", None):
+    if spec["package"] in develPkgs and getattr(syncHelper, "writeStore", False):
```
Comment (ktf) [inline alibuild_helpers/build.py:730]: Why changing this here? If it is not strictly necessary, can you move it to a separate PR (which can go in immediately)?

Comment (TimoWilken) [reply alibuild_helpers/build.py:730]: OK, will do.

File: alibuild_helpers/build.py (hunk 9 above)
Comment (ktf) [reply alibuild_helpers/build.py:833]: Yes, please. We can merge that PR immediately. Given the PR is about a quite critical change, I would like it to be distilled to its minimum.

**Comment by TimoWilken** [comment]
One minute, I just realised I reverted a bit too much.

File: alibuild_helpers/build.py (hunk 7 above)
Comment (TimoWilken) [reply alibuild_helpers/build.py]: Done.

File: alibuild_helpers/build.py (hunk 10 above)
Comment (TimoWilken) [reply alibuild_helpers/build.py]: Reverted or now; I'll include it in the separate cleanup PR.

File: alibuild_helpers/build.py (hunk 9 above)
Comment (TimoWilken) [reply alibuild_helpers/build.py]: Reverted.

File: alibuild_helpers/build.py (hunk 8 above)
Comment (TimoWilken) [reply alibuild_helpers/build.py]: Reverted; I'll include it in the cleanup PR.
